java/ql/lib/semmle/code/java/security/InsecureTrustManagerQuery.qll

```diff
@@ -0,0 +1,20 @@
+/** Provides taint tracking configurations to be used in Trust Manager queries. */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.security.Encryption
+import semmle.code.java.security.InsecureTrustManager
+
+/**
+ * A configuration to model the flow of an insecure `TrustManager`
+ * to the initialization of an SSL context.
+ */
+class InsecureTrustManagerConfiguration extends TaintTracking::Configuration {
```

Comment:
Can this be changed to `DataFlow`? I've used taint tracking due to limitations of data flow at the time.
See this comment from `@aschackmull`:
https://github.com/github/codeql/pull/4879#discussion_r617427665

Reply:
`DataFlow` could be used, but then we would need to allow `Array` implicit reads in the sinks (and potentially in the additional flow steps). Something like:

```ql
override predicate allowImplicitRead(DataFlow::Node node, DataFlow::Content c) {
  (this.isSink(node) or this.isAdditionalFlowStep(node, _)) and
  node.getType() instanceof Array and
  c instanceof DataFlow::ArrayContent
}
```

(which is actually part of the default implicit reads in `TaintTracking`).

We could even restrict `node` to be an array of `TrustManager`s, since that's what [the sink is expecting](https://docs.oracle.com/javase/7/docs/api/javax/net/ssl/SSLContext.html#init(javax.net.ssl.KeyManager[],%20javax.net.ssl.TrustManager[],%20java.security.SecureRandom)).

Of course, the trade-off would be that, if for some reason the insecure `TrustManager` temporary goes through more complex data structures, e.g. a `Map`, we'd lose flows that a `TaintTracking::Configuration` would catch, but that doesn't sound like something that happens often.

@aschackmull what do you think is more appropriate in this case?

Reply:
I did this in 967308f.